sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureFileSystem.java

```diff
@@ -0,0 +1,259 @@
+ * file store is represented by a container in the storage account. Each container has one root directory.
+ *
+ * Closing the file system will not block on outstanding operations. Any operations in progress will be allowed to
+ * terminate naturally after the file system is closed, though no further operations may be started after the parent
+ * file system is closed.
+ * {@inheritDoc}
+ */
+public final class AzureFileSystem extends FileSystem {
+    // Configuration constants for blob clients.
+    public static final String AZURE_STORAGE_ACCOUNT_KEY = "AzureStorageAccountKey";
+    public static final String AZURE_STORAGE_SAS_TOKEN = "AzureStorageSasToken";
+    public static final String AZURE_STORAGE_HTTP_LOG_DETAIL_LEVEL = "AzureStorageHttpLogDetailLevel";
+    public static final String AZURE_STORAGE_MAX_TRIES = "AzureStorageMaxTries";
+    public static final String AZURE_STORAGE_TRY_TIMEOUT = "AzureStorageTryTimeout";
+    public static final String AZURE_STORAGE_RETRY_DELAY_IN_MS = "AzureStorageRetryDelayInMs";
+    public static final String AZURE_STORAGE_MAX_RETRY_DELAY_IN_MS = "AzureStorageMaxRetryDelayInMs";
+    public static final String AZURE_STORAGE_RETRY_POLICY_TYPE = "AzureStorageRetryPolicyType";
+    public static final String AZURE_STORAGE_SECONDARY_HOST = "AzureStorageSecondaryHost";
+    public static final String AZURE_STORAGE_BLOCK_SIZE = "AzureStorageBlockSize";
+    public static final String AZURE_STORAGE_DOWNLOAD_RESUME_RETRIES = "AzureStorageDownloadResumeRetries";
+    public static final String AZURE_STORAGE_USE_HTTPS = "AzureStorageUseHttps";
+    public static final String AZURE_STORAGE_HTTP_CLIENT = "AzureStorageHttpClient"; // undocumented; for test.
+
+    public static final String AZURE_STORAGE_FILE_STORES = "AzureStorageFileStores";
+
+    private static final String AZURE_STORAGE_ENDPOINT_TEMPLATE = "%s://%s.blob.core.windows.net";
+
+    private final AzureFileSystemProvider parentFileSystemProvider;
+    private final BlobServiceClient blobServiceClient;
+    private final Integer blockSize;
+    private final Integer downloadResumeRetries;
+    private final Map<String, FileStore> fileStores;
+    private boolean closed;
+
+    AzureFileSystem(AzureFileSystemProvider parentFileSystemProvider, String accountName, Map<String, ?> config)
+            throws IOException {
+        // A FileSystem should only ever be instantiated by a provider.
+        if (Objects.isNull(parentFileSystemProvider)) {
+            throw new IllegalArgumentException("AzureFileSystem cannot be instantiated without a parent " +
+                "FileSystemProvider");
+        }
+        this.parentFileSystemProvider = parentFileSystemProvider;
+
+        // Read configurations and build client.
+        try {
+            this.blobServiceClient = this.buildBlobServiceClient(accountName, config);
+            this.blockSize = (Integer) config.get(AZURE_STORAGE_BLOCK_SIZE);
```

Comment:
Is this block size for upload? Do we want to clarify that in the name?

Reply:
It's the block size we pass to the ParallelTransferOptions. I guess in this case it would always be upload, so I'll update the name.